model/version.go

```diff
@@ -356,6 +356,39 @@ func VersionGetHistory(versionId string, N int) ([]Version, error) {
 	return versions, nil
 }
 
+type MainlineCommitVersionOptions struct {
+	Limit int `json:"limit"`
+	Skip  int `json:"skip"`
+}
+
+func GetMainlineCommitVersionsWithOptions(projectName string, opts MainlineCommitVersionOptions) ([]Version, error) {
+	projectId, err := GetIdForProject(projectName)
+	if err != nil {
+		return nil, err
+	}
+	match := bson.M{
+		VersionIdentifierKey: projectId,
+	}
+	pipeline := []bson.M{bson.M{"$match": match}}
+	pipeline = append(pipeline, bson.M{"$sort": bson.M{VersionRevisionOrderNumberKey: -1}})
+
+	limit := defaultVersionLimit
+	if opts.Limit != 0 {
+		limit = opts.Limit
+	}
+	pipeline = append(pipeline, bson.M{"$limit": limit})
+
+	res := []Version{}
+
+	if err := db.Aggregate(VersionCollection, pipeline, &res); err != nil {
```

Comment:
So this was intentional because of my last checkbox in the description. I'm not sure how to actually perform pagination for this type of query i figured an aggregation would offer the most flexibility. But i essentially want a query where the limit only applies to the number of matched values. but the values that do not match should still be included in the response.

Reply:
Sorry i didn’t read your comment here well enough! Yeah so this isn't going to do what you want yet. And an issue I'm seeing is that we don't store if versions are activated for older versions, so if someone goes back far enough in the waterfall, how will we handle that? Do we could Activated = nil versions towards the limit or not? 

I think there's not a way to apply limit to only one field while not limiting the results returned. Some ideas:
1. Use two queries: the first query can get the activated versions, and then the second query can get the not activated versions that have a revision order number less than the last activated version (would need to factor in nil activations, maybe by adding them to the right place in the activated versions list and then dropping unactivated versions that are now "out of range").

> 2. Use SetBatchSize() for pagination. You can initially set the batchsize to the limit you want, and then iterate with `.Next()` if (after checking activation statuses) you need more. These two ideas might actually work well together, if you make query1 "Activated or nil" and then iterate again if nil versions end up being not-activated.

In both cases, it might make sense to move the "check if activated is set" logic here. And these ideas are pretty hacky; @john-m-liu is more of a query expert and may be able to think of an aggregation that makes sense.

Reply:
I actually don't have much to add, I think Annie is spot on (I also don't think the 2 query thing is hacky at all, it's how I would do it). The only thing is that the match part is missing a filter on requester and potentially other stuff, so you're currently returning both patches and commits. https://github.com/evergreen-ci/evergreen/blob/ede384502d637976b7e86f0982a858bfbc0a9b03/service/waterfall.go#L200 is the query on the old waterfall if that's helpful

Reply:
https://jira.mongodb.org/browse/EVG-14615 Created this ticket to cover the backfilling logic. This will be done after this pr and before any one uses this page and encounters a version without this field

Reply:
Thanks I got a working approach using the 2 queries for activated and unactivated versions and merging them and sorting by order number. 
We paginate based off of the order number. So if its provided it will fetch all the versions that came before the order number

Reply:
I spoke to @bsamek  about the backfilling logic. It might be better not to do it. Since by the time users start using this page it will already be several months after this field was introduced. So there will be a low likely hood of someone actually encountering a version without this field.